### ISSUE TITLE
Allow syntax highlighter extensions in document header

### ISSUE THREAD

**Issue by pilhuhn** [asciidoctor/asciidoctor#1913]
Syntax highlighters like pygments have a lot of languages build in.
Suppose I want now write an article about a DSL that looks a bit like Ruby (Treetop grammar in this case).

I could go

```
[source, ruby]
----
grammar MyDSL
  rule eval_boolean
    space ( 'and' / 'or' ) space
  end
end
----

```

which then colors the `end` correctly, but does not know about the `rule` nor `grammar`. 

With this feature request it would be possible to tell the highlighter somehow in the document header that `rule` and `grammar` are to be treated like a `if` keyword in Ruby.

```
= My First DSL
:highlighter-ext: ...<specfic rule>
```

**Comment by mojavelinux** [comment]
Putting this type of information in the header would not be in the spirit of metadata. It's too coupled. This is where an extension library would come in.

Both Pygments and Rouge allow for additional languages to be added by putting language definitions on the load path. Obviously, this is something we need to document more clearly. I recall efforts by other users to do something like this in the past, so I know that it can be done.

That information can be found in #1382. I'll mark this as a duplicate of that issue.

**Comment by pilhuhn** [comment]
Let me illustrate my use case (which I unfortunately did not do while opening the issue).
I am writing a document about Treetop and want to highlight the additional keywords (on top of Ruby). I am editing this in Intellij/RubyMine+ Asciidoc plugin. To export I open it in AsciidocFX. And then later I put it on git(hub/lab) and perhaps post it on Hawkular.org in the blog done by JBake.
Perhaps I even have a collaborator on git(hub/lab).

With an extension library I have to dig into all the places where the highlighter library is hidden and install my rules. I have to potentially even run irb to refresh some asciidoctor cache as described in #1382. And all the other people involved have to do this as well.

Your comment above made me thing though about a level of indirection.
Instead of

```
= My First DSL
:highlighter-ext: ...<specfic rule>

```

It should better be
```
= My First DSL
:extension: { 'pygments' : ...<pygments specfic rule> }
```

This way the namespace is not polluted by XX-extension per highlighter/library, but rather asciidoctor would pass the block to the library that is listed as first argument (pygments in above example).

**Comment by mojavelinux** [comment]
Thanks for the additional info. I understand the nuances of your scenario better. However, this is still way out of scope for core. But you could certainly accomplish what you want to accomplish using a Treeprocessor (which can visit any block). You could even introduce your own document-level attributes that the Treeprocessor will use. So this is all very possible. It's just no something I think core should be touching.

**Comment by mojavelinux** [comment]
As of Asciidoctor 2.0, the syntax highlighter is an extension point. That means it's possible to take over how syntax highlighting is done, either by extending a provided implementation or writing one from scratch.

Asciidoctor 2.0 also provides Rouge integration out of the box. This is significant because it's easy to contribute or override a lexer in Rouge. So you won't even have to touch Asciidoctor behavior in that case.

In short, the tools are now there to do this. I don't think any additional changes to Asciidoctor are warranted.

For information about the syntax highlighter API, see https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/syntax_highlighter.rb
